### PR TITLE
feat(provider): Add africastalking provider

### LINKED ISSUE #1330: 🚀 Feature: Africastalking SMS Provider
### 🔖 Feature description

SMS Provider for Novu

### 🎤 Why is this feature needed ?

From AT landing page, "Powering Communications Solutions Across Africa"
As one of the leading SMS provider in Africa, Using Novu for African focused application will be easier than ever

### ✌️ How do you aim to achieve this?

building an AT SMS Provider for Novu

### 🔄️ Additional Information

_No response_

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Contributing Guidelines](https://github.com/novuhq/novu/blob/main/CONTRIBUTING.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by peoray**
### What change does this PR introduce?
Add Africastalking provider

Explain here the changes your PR introduces and text to help us understand the context of this change.

An updated version of #3102. Resolved all errors and make use of new changes introduced in the codebase.

### Why was this change needed?

Closes #1330

### CODE REVIEW

File: libs/shared/src/consts/providers/channels/sms.ts [hunk 1]
```diff
@@ -153,4 +154,12 @@ export const smsProviders: IProviderConfig[] = [
     docReference: 'https://developers.termii.com/authentication',
     logoFileName: { light: 'termii.png', dark: 'termii.png' },
   },
+  {
+    id: SmsProviderIdEnum.Africastalking,
```
Comment (p-fernandez) [inline libs/shared/src/consts/providers/channels/sms.ts:158]: Seeing the branded name I'd go for `AfricasTalking` / `africasTalking` capitalisation everywhere as needed.

File: libs/shared/src/consts/providers/channels/sms.ts [hunk 2]
```diff
@@ -153,4 +154,12 @@ export const smsProviders: IProviderConfig[] = [
...
+    displayName: 'Africastalking',
```
Comment (p-fernandez) [inline libs/shared/src/consts/providers/channels/sms.ts:159]:
```suggestion
    displayName: `Africa's Talking`,
```

File: libs/shared/src/consts/providers/provider.enum.ts [hunk 3]
```diff
@@ -63,6 +63,7 @@ export enum SmsProviderIdEnum {
   Kannel = 'kannel',
   Maqsam = 'maqsam',
   Termii = 'termii',
+  Africastalking = 'africastalking',
```
Comment (p-fernandez) [inline libs/shared/src/consts/providers/provider.enum.ts:66]:
```suggestion
  AfricasTalking = 'africas-talking',
```

File: packages/application-generic/package.json [hunk 4]
```diff
@@ -87,6 +87,7 @@
     "@novu/termii": "^0.13.0",
     "@novu/testing": "^0.13.0",
     "@novu/twilio": "^0.13.0",
+    "@novu/africastalking": "^0.13.0",
```
Comment (p-fernandez) [inline packages/application-generic/package.json:90]: Sorry for being annoying but let's rename the folder and packaging name to `africas-talking` due the company's branding. 🙏🏻

File: providers/africastalking/README.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+## Usage
+
+```javascript
+import { AfricastalkingSmsProvider } from '@novu/africastalking';
+
+ const provider = new AfricastalkingSmsProvider({
+    apiKey: process.env.AFRICASTALKING_API_KEY,
+    username: process.env.AFRICASTALKING_USERNAME,
+    from: process.env.AFRICASTALKING_FROM
+  });
```
Comment (p-fernandez) [inline providers/africastalking/README.md:14]:
```suggestion
# Novu Africa's Talking Provider

A Africa's Talking SMS provider library for [@novu/node](https://github.com/novuhq/novu)

## Usage

```javascript
import { AfricasTalkingSmsProvider } from '@novu/africas-talking';

 const provider = new AfricasTalkingSmsProvider({
    apiKey: process.env.AFRICAS_TALKING_API_KEY,
    username: process.env.AFRICAS_TALKING_USERNAME,
    from: process.env.AFRICAS_TALKING_FROM
  });
```
Besides suggestions for the content, fixes for my suggested code naming based on the branding.

File: providers/africastalking/jest.config.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
+module.exports = {
+  preset: 'ts-jest',
+  testEnvironment: 'node',
+  moduleNameMapper: {
+    axios: 'axios/dist/node/axios.cjs',
+  },
```
Comment (p-fernandez) [inline providers/africastalking/jest.config.js:7]: 👏🏻

File: providers/africastalking/package.json [hunk 7]
```diff
@@ -0,0 +1,83 @@
+{
+  "name": "@novu/africastalking",
+  "version": "0.13.0",
+  "description": "A africastalking wrapper for novu",
```
Comment (p-fernandez) [inline providers/africastalking/package.json:4]:
```suggestion
  "description": "A Africa's Talking wrapper for Novu",
```

File: providers/africastalking/src/lib/africastalking.provider.spec.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import { AfricastalkingSmsProvider } from './africastalking.provider';
+
+test('should trigger africastalking library correctly', async () => {
```
Comment (p-fernandez) [inline providers/africastalking/src/lib/africastalking.provider.spec.ts:3]:
```suggestion
test(`should trigger Africa's Talking library correctly`, async () => {
```
Suggestion for the trickier situations due branding apostrophe.

File: providers/africastalking/src/lib/africastalking.provider.spec.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  const provider = new AfricastalkingSmsProvider({
+    apiKey: 'b664b089f04b72c56ac3b0a8ffbb6f3d18a82eb40c29d17b49b84433439fb127',
```
Comment (p-fernandez) [inline providers/africastalking/src/lib/africastalking.provider.spec.ts:5]: Tell me it is not a real apiKey... 😅

File: providers/africastalking/src/lib/africastalking.provider.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  ): Promise<ISendMessageSuccessResponse> {
+    const response = await this.africastalkingClient.send({
+      from: this.config.from,
+      to: options.to,
+      message: options.content,
+    });
+
+    return {
+      id: response?.SMSMessageData?.Recipients[0]?.messageId,
+      date: new Date().toISOString(),
```
Comment (p-fernandez) [inline providers/africastalking/src/lib/africastalking.provider.ts:38]: Do we get any timestamp in the response of the call from the provider? If so, better use it that autogenerate it.

**Review by p-fernandez — CHANGES_REQUESTED** [review]
Thank you for recreating the PR. Mostly changes based on the branding name of the provider. 🙏🏻

File: providers/africastalking/src/lib/africastalking.provider.spec.ts (hunk 9 above)
Comment (peoray) [reply providers/africastalking/src/lib/africastalking.provider.spec.ts:5]: no it's not 😁

**Comment by peoray** [comment]
@p-fernandez resolved

File: .vscode/settings.json [hunk 11]
```diff
@@ -1,20 +1,23 @@
...
+          "editor.defaultFormatter": "dbaeumer.vscode-eslint",
+            "editor.formatOnSave": true,
+              "eslint.format.enable": true,
+                "editor.tabSize": 2,
+                  "files.insertFinalNewline": true,
+                    "files.trimFinalNewlines": true,
+                      "css.validate": false,
+                        "editor.wordWrap": "wordWrapColumn",
+                          "editor.wordWrapColumn": 120,
+                            "editor.codeActionsOnSave": {
```
Comment (p-fernandez) [inline .vscode/settings.json:15]: I don't think this indentation is right.

File: packages/application-generic/src/factories/sms/handlers/africas-talking.handler.ts [hunk 12]
```diff
@@ -0,0 +1,24 @@
+import { ChannelTypeEnum } from '@novu/shared';
+import { AfricasTalkingSmsProvider } from '@novu/africas-talking';
+import { ICredentials } from '@novu/dal';
+import { BaseSmsHandler } from './base.handler';
+
+export class AfricasTalkingSmsHandler extends BaseSmsHandler {
+  constructor() {
+    super('africastalking', ChannelTypeEnum.SMS);
```
Comment (p-fernandez) [inline packages/application-generic/src/factories/sms/handlers/africas-talking.handler.ts:8]:
```suggestion
    super('africas-talking', ChannelTypeEnum.SMS);
```
Based on the enum value set.

File: providers/africas-talking/src/lib/africas-talking.provider.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
+import {
+  ChannelTypeEnum,
+  ISendMessageSuccessResponse,
+  ISmsOptions,
+  ISmsProvider,
+} from '@novu/stateless';
+import AfricasTalking from 'africasTalking';
+
+export class AfricasTalkingSmsProvider implements ISmsProvider {
+  id: 'africastalking';
```
Comment (p-fernandez) [inline providers/africas-talking/src/lib/africas-talking.provider.ts:10]:
```suggestion
  id: 'africas-talking';
```

**Review by p-fernandez — APPROVED** [review]
2 small things and is done.
Also as we published the Novu version 0.14.0 you will need to upgrade the version number to 0.14.0. Also, very important the version of `@novu/stateless` dependency to this newest version.

**Comment by peoray** [comment]
@p-fernandez resolved

**Comment by p-fernandez** [comment]
> @p-fernandez resolved

We need a `pnpm install` to fix the lock file conflicts. 🙏🏻

**Comment by peoray** [comment]
@p-fernandez Not sure what you mean, but I went ahead to pull in the latest changes and did a `pnpm install` as well as `npm run setup:project`

**Comment by p-fernandez** [comment]
> @p-fernandez Not sure what you mean, but I went ahead to pull in the latest changes and did a `pnpm install` as well as `npm run setup:project`

Basically what you did. 😄

File: providers/africas-talking/src/lib/africas-talking.provider.ts [hunk 14]
```diff
@@ -0,0 +1,41 @@
+import {
+  ChannelTypeEnum,
+  ISendMessageSuccessResponse,
+  ISmsOptions,
+  ISmsProvider,
+} from '@novu/stateless';
+import AfricasTalking from 'africasTalking';
```
Comment (p-fernandez) [inline providers/africas-talking/src/lib/africas-talking.provider.ts:7]:
```suggestion
import AfricasTalking from 'africastalking';
```

File: providers/africas-talking/src/lib/africas-talking.provider.spec.ts [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+test(`should trigger Africa's Talking library correctly`, async () => {
+  const provider = new AfricasTalkingSmsProvider({
+    apiKey: 'b664b089f04b72c56ac3b0a8ffbb6f3d18a82eb40c29d17b49b84433439fb127',
+    username: 'sandbox',
+    from: '1234',
+  });
+
+  const spy = jest
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
```
Comment (p-fernandez) [inline providers/africas-talking/src/lib/africas-talking.provider.spec.ts:12]:
```suggestion
    // @ts-expect-error
```

File: providers/africas-talking/src/lib/africas-talking.provider.spec.ts [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+  await provider.sendMessage({
+    to: '+2347063317344',
+    content: 'SMS Content',
+  });
+
+  expect(spy).toHaveBeenCalled();
+  expect(spy).toHaveBeenCalledWith({
+    to: '+2347063317344',
+    from: '2585',
```
Comment (p-fernandez) [inline providers/africas-talking/src/lib/africas-talking.provider.spec.ts:29]:
```suggestion
    from: '1234',
```

**Pull request merged**